### PR TITLE
allow delegated authorization to have privileged groups

### PR DESCRIPTION
**Pull request by deads2k**
In kube, `system:masters` is a special group that has full API powers.  We use this group to be able to have kube-apiservers loopback with full power and to allow management of RBAC using a client cert with that group.

Since DelegatedAuthorization is intended to be consistent with kube recommendations, this updates the default value to include an authorizer that lists `system:masters` as having full rights. This means that system:masters users will not have to have a remote authz call which improves performance and makes debugging very broken conditions possible.

@smarterclayton 
@kubernetes/sig-api-machinery-pr-reviews @kubernetes/sig-auth-pr-reviews 

```release-note
delegated authorization can now allow unrestricted access for `system:masters` like the main kube-apiserver
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/authorization.go [hunk 1]
```diff
@@ -56,16 +57,28 @@ type DelegatingAuthorizationOptions struct {
...
 }
 
 func NewDelegatingAuthorizationOptions() *DelegatingAuthorizationOptions {
 	return &DelegatingAuthorizationOptions{
 		// very low for responsiveness, but high enough to handle storms
 		AllowCacheTTL: 10 * time.Second,
 		DenyCacheTTL:  10 * time.Second,
+
+		// in a kube cluster, system:masters has full power.  Use this as the default which could be later changed.
+		AlwaysAllowGroups: []string{user.SystemPrivilegedGroup},
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/authorization.go]: I think I'm on board with defaulting to a value that allows coasting, but I expected a flag wired to this to surface it and allow control

Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/options/authorization.go]: > I think I'm on board with defaulting to a value that allows coasting, but I expected a flag wired to this to surface it and allow control

This seems like a setting that individual binaries can choose to control how they like.  I think a very small minority (if any) would want to allow an admin to control it and those binaries could choose to wire the value to whatever flag they like.

**Comment by mikedanese** [comment]
I'm worried that:
* This is a leak of an RBAC specific detail into the scope of general authorization
* Bypasses short-circuit deny authorizers

**Comment by liggitt** [comment]
> * This is a leak of an RBAC specific detail into the scope of general authorization
> * Bypasses short-circuit deny authorizers

the same superuser exists in the kube-apiserver as the pinned first authorizer in the chain, independent of RBAC and prior to any configured authorizers

**Comment by mikedanese** [comment]
Ahh... loopback stuff... I have similar concerns with that flow. While we couldn't make the same guarantees around preventing misconfiguration, I would strongly prefer if the loopback authorizer was appended to the authorizer chain rather than prepended. The usecase was outlined here https://github.com/kubernetes/kubernetes/issues/51862

**Comment by mikedanese** [comment]
I'm primarily concerned with introducing well known user/group strings that are capable of unilateral bypassing authorization when operators generally have little control over the users and groups of credentials (namely those provisioned by the certificates API).

> I would strongly prefer if the loopback authorizer was appended to the authorizer chain rather than prepended.

Alternatively, we could enforce that the loopback authorizer could only authorize request from loopback addresses.

**Comment by deads2k** [comment]
I can make it optional, but I think that prepending is the correct default choice. This is already what the kube-apiserver does, so this is purely an optimization on asking.  Again, a user of this API can wire or disable this in any way that they like and removing the default makes it purely opt-in.

**Comment by deads2k** [comment]
Default removed.  If someone requests it, then it will function as desired.

**Comment by liggitt** [comment]
/kind feature
/priority important-soon
/lgtm
/approve

needs release note

**Pull request merged**